docs/architecture/adr-071-bank-v2.md

```diff
@@ -0,0 +1,183 @@
+
+In addition to the above, the bank module is currently too rigid and handles too many tasks, so this proposal aims to streamline the module by focusing on core functions `Send`, `Mint`, and `Burn`.
+
+Currently, the module is split accross different keepers with scattered and duplicates functionalities (with 4 send functions for instance).
+
+Additionally, the integration of the token factory into the bank module allows for standardization, and better integration within the core modules.
+
+This rewrite will reduce complexity and enhance the efficiency and UX of the bank module.
+
+## Context
+
+The current implementation of the bank module is characterised by its handling of a broad array of functions, leading to significant complexity in using and extending the bank module. 
+
+These issues have underscored the need for a refactoring strategy that simplifies the module’s architecture and focuses on its most essential operations.
+
+Additionally, there is an overlap in functionality with a Token Factory module, which could be integrated to streamline oper.
+
+## Decision
+
+**Permission Tightening**: Access to the module can be restricted to selected denominations only, ensuring that it operates within designated boundaries and does not exceed its intended scope. Currently, the permissions allow all denoms, so this should be changed. Send restrictions functionality will be maintained.
+
+**Simplification of Logic**: The bank module will focus on core functionalities `Send`, `Mint`, and `Burn`. This refinement aims to streamline the architecture, enhancing both maintainability and performance.
+
+**Integration of Token Factory**: The Token Factory will be merged into the bank module. This consolidation of related functionalities aims to reduce redundancy and enhance coherence within the system. Migrations functions will be provided for migrating from Osmosis' Token Factory module to bank/v2.
+
+**Legacy Support**: A legacy wrapper will be implemented to ensure compatibility with about 90% of existing functions. This measure will facilitate a smooth transition while keeping older systems functional.
+
+**Callback Functions**: We propose to integrate callback functions directly to the x/bank module, allowing for customisable behaviour during minting, burning and transferring operations without complicating statemanagement. It also helps overwrite the default accounting mechanism by allowing to overwrite the balance method (this is particularly useful for rebasing tokens).
+
+Implementation of Callback Functions:
+
+```go
+type BankKeeper struct {
+    OnMint func(ctx context.Context, addr []byte, amount sdk.Coin) error
+    OnBurn func(ctx context.Context, addr []byte, amount sdk.Coin) error
+    OnTransfer func(ctx context.Context, fromAddr, toAddr []byte, amount sdk.Coin) error
+    OnBalanceOf func(ctx context.Context, addr []byte, denom string) (sdk.Coins, error)
+}
```

Comment:
What we discussed was these callbacks being per denom and each denom being (for now) an x/account instance that handles the callback message type. Each callback is optional.

Also, Mint and Burn callbacks could be excluded if we make it so that the denom admin is the only account who can mint or burn. This would of course mean getting rid of the bank keeper Mint and Burn functions, but I think those and the other unsafe keeper methods should go anyway.